src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+    // set central visited node to visted
+    visited[0] = true;
+
+    // call recursive helper fuction to get final Answer
+    this.finalAnswer = solverHelper(currPos, numNodes, currentTime, count, 
+        cost, ans, visited);
+  }
+
+  /**
+   * Recursive helper function that is used to solve computational part of
+   * algorithm. Called in global `solver()` method
+   *
+   * @param currPos represents the placeId of where the alg currently is
+   * @param numNodes total number of locations to visit (includes center)
+   * @param currentTime time if path at this point were taken 
+   * @param count number of pois visited during this path
+   * @param cost Tuple of current path (time cost, List of pois)
+   * @param ans the current best solution found
+   * @param array of which pois have been visited by this path
+   */
+  public Tuple solverHelper(int currPos, int numNodes, LocalTime currentTime, 
+      int count, Tuple cost, Tuple ans, boolean[] visited) {
+    /**
+     * If last node is reached and shares an edge w/ start node
+     * Calculate min of cost and currAns
+     * return and keep traversing the graph/matrix
+     */
+    if ((count == numNodes-1) && timeMatrix[currPos][0] > 0) {
+      if (ans.getCurrAns() < cost.getCurrAns() + timeMatrix[currPos][0]) {
+        return ans;
+      } else {
+        cost.incCurrAns(timeMatrix[currPos][0]);
+        return cost;
+      }
+    }
+
+    /**
+     * Backtracking Step: loop through the time Matrix from currPos, increase
+     * count by 1 and update (int) and (List) value of cost
+     */
+    for (int i = 0; i < numNodes; i++) {
+      // set open boolean and time to add to currenTime
+      boolean isOpen = (boolean) checkIsOpenSetTime(currentTime, i,
+          currPos).get(0);
+      int timeToAdd = (int) checkIsOpenSetTime(currentTime, i,
+          currPos).get(1);
+      
+      // if node is unvisited and greater than 0, i.e. not the same node
+      if (!visited[i] && timeMatrix[currPos][i] > 0 && isOpen) {
+        // mark ith node as visited
+        visited[i] = true;
+
+        // update cost
+        LocalTime timePostUpdate = currentTime.plusMinutes((long) timeToAdd);
+        List<Integer> costPath = cost.getCurrPath();
+        List<Integer> updatePath = new ArrayList<>();
+        updatePath.addAll(costPath);
+        updatePath.add(i);
+        int newTupleCost = cost.getCurrAns() + timeToAdd - HOUR;
+        Tuple updateTuple = new Tuple(newTupleCost, updatePath);
+
+        // recursive call
+        ans = solverHelper(i, numNodes, timePostUpdate, count + 1, 
+            updateTuple, ans, visited);
+
+        // mark ith node as unvisited
+        visited[i] = false;
+      }
+    }
+    return ans;
+  }
+
+  /**
+   * Check if a location is or will be open. Possibilities:
+   * 1. Arrival before open time 
+   * 2. Arrival during open hours
+   * 3. Arrival after closed 
+   * 4. Open Hours are not available
+   * 
+   * @return List<Object> of length 2 
```

Comment:
change requested: mixing types in List is usually discouraged.

A more typical way to do this would be to just return an int but have a "sentinel value" to indicate case 3. Eg: either return the time added (some number > 0) or returns -1 if arriving after close. This makes it easier for the code that calls this method because it doesn't have to deal with casting elements from the returned List<Object>.

Reply:
Right! This is much better: it felt weird/wrong to have a list of different objects.